syntax = "proto3";

package libcore;
option go_package = "libcore/nekoray_rpc";

service LibcoreService {
  rpc Exit(EmptyReq) returns (EmptyResp) {}
  rpc KeepAlive(EmptyReq) returns (EmptyResp) {}
  //
  rpc Start(LoadConfigReq) returns (ErrorResp) {}
  rpc Stop(EmptyReq) returns (ErrorResp) {}
  rpc Test(TestReq) returns (TestResp) {}
  rpc QueryStats(QueryStatsReq) returns (QueryStatsResp) {}
}

message EmptyReq {}

message EmptyResp {}

message ErrorResp {
  string error = 1;
}

message LoadConfigReq {
  string coreConfig = 1;
}

enum TestMode {
  TcpPing = 0;
  UrlTest = 1;
}

message TestReq {
  TestMode mode = 1;
  int32 timeout = 6;
  // TcpPing
  string address = 2;
  // UrlTest
  LoadConfigReq config = 3;
  string inbound = 4;
  string url = 5;
}

message TestResp {
  string error = 1;
  int32 ms = 2;
}

message QueryStatsReq{
  string tag = 1;
  string direct = 2;
}

message QueryStatsResp{
  int64 traffic = 1;
}
